appsScriptAuth.js

```diff
@@ -0,0 +1,166 @@
+ * enable Google Cloud Logging to retain them, which I don't feel like doing right now :)
+ */
+
+// webapp deployment entry point
+// expects a GET request with a parameter "user" which is a single email address or an array of the same
+// https://developers.google.com/apps-script/guides/web
+function doGet(e) {
+  var authorized = false;
+  if ('user' in e.parameters) {
+    authorized = isUserAuthorizedForChive([e.parameter['user']])
+  } else if ('user[]' in e.parameters) {
+    authorized = isUserAuthorizedForChive(e.parameters['user[]'])
+  }
+  return ContentService.createTextOutput(JSON.stringify({authorized: authorized})).setMimeType(ContentService.MimeType.JSON);
+}
+
+// takes a list of email addresses, and determines whether any of them are authorized to access a
+// hard-coded Google Drive folder ID.
+function isUserAuthorizedForChive(userEmailList) {
+  var authList = getEmailsInDriveFolder("PUT_YOUR_DRIVE_FOLDER_ID_HERE")
+  var expandedAuthList = expandEmails(authList)
+  var normalizedAuthList = normalizeEmails(expandedAuthList)
+
+  var normalizedUserList = normalizeEmails(userEmailList)
+  
+  Logger.log(normalizedAuthList);
+  Logger.log(normalizedUserList)
+
+  var authorized = false;
+  for (userEmail of normalizedUserList) {
+    if (normalizedAuthList.includes(userEmail)) {
+      authorized = true;
+      break;
+    }
+  }
+  return authorized;
+}
+
+// returns a list of email addresses that are viewers or editors of a specific Google Drive folder
+// note: account that script runs as must have editor access to the folder to get permissions with this API
+function getEmailsInDriveFolder(folderid) {
+  var drive = DriveApp.getFolderById(folderid)
+  var emails = [];
+  for (let viewer of drive.getViewers()) {
+    emails.push(viewer.getEmail())
+  }
+  for (let editor of drive.getEditors()) {
+    emails.push(editor.getEmail())
+  }
+  return emails;
+}
+
+// returns a list of email addresses that are the member of a Google Group (based on its x@googlegroups.com handle)
+// note: account that script runs as must be a member of the group to get the member list with this API
+function getEmailsInGoogleGroup(grouphandle) {
+  try {
+    var group = GroupsApp.getGroupByEmail(grouphandle);
+    var emails = [];
+    var users = group.getUsers();
+    for (let user of users) {
+      emails.push(user.getEmail())
+    }
+    return emails;
+  } catch (e) {
+    // just in case GroupsApp call fails
+    Logger.log(e)
+    return [];
+  }
+}
+
+// expand any Google Groups handles into their member lists
+// do this in a single loop; this means that if there are ever recursive groups, they will not be expanded
+// [just don't make recursive groups please]
+function expandEmails(emails) {
+  var output = []
+  for (let email of emails) {
+    [user, domain] = email.split("@")
+    domain = domain.toLowerCase()
+    if (domain == "googlegroups.com") {
```

Comment:
The argument to this function is the list of emails authorized to access the Chive, so it ends up looking like
```
[alum@gmail.com, ourmaingroup@googlegroups.com, ourtechgroup@googlegroups.com, serviceaccount@serviceaccount.com, topher@gmail.com]
```
So, the purpose of this is to automatically expand the groups we give access to the Chive. Most people aren't individually enumerated on the permissions list.

Reply:
Or maybe you were asking about the comment about recursive groups? That doesn't happen with our current setup, but theoretically it could. https://support.google.com/a/answer/167100